In [8]:
import pandas as pd

In [9]:
schooldata_df = pd.read_csv("schools_complete.csv") 
studentdata_df = pd.read_csv("students_complete.csv") 

# Read school data into pandas
schooldata_df
schooldata_df.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [10]:
# Read student data into pandas 
studentdata_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [11]:
# Combine the data into a single dataset
school_data_complete = pd.merge(studentdata_df, schooldata_df, how ="left", on=["school_name", "school_name"])
school_data_complete

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[39170 rows x 11 columns]

In [12]:
#Calculate total number of schools
totalschools_df = schooldata_df.count()
totalschools_df

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [13]:
#Calculate total number of students  
totalstudents_df = studentdata_df.count()
totalstudents_df

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [14]:
#Calculate the total budget
totalbudget_df = schooldata_df["budget"].sum()
totalbudget_df

24649428

In [15]:
#calculate average reading mean
reading_mean_df = studentdata_df["reading_score"].mean()
reading_mean_df

81.87784018381414

In [16]:
#calculate average math mean
math_mean_df = studentdata_df["math_score"].mean()
math_mean_df

78.98537145774827

In [17]:
#Calculate overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2
overall_avg_score_df = reading_mean_df + math_mean_df 
overall_avg_score_df/2


80.43160582078121

In [18]:
#Calculate the percentage of students with a passing math score (70 or greater)
    
studentdata_df["#passing_math"] = studentdata_df["math_score"] >= 70
studentdata_df["#passing_reading"] = studentdata_df["reading_score"] >= 70
percent_passing_math = ((studentdata_df["#passing_math"]).mean())*100
percent_passing_math

74.9808526933878

In [19]:
#Calculate the percentage of students with a passing reading score (70 or greater)

studentdata_df["#passing_math"] = studentdata_df["math_score"] >= 70
studentdata_df["#passing_reading"] = studentdata_df["reading_score"] >= 70
percent_passing_reading = ((studentdata_df["#passing_reading"]).mean())*100
percent_passing_reading

85.80546336482001

In [20]:
# Creating the DataFrame 
district_summary_df = pd.DataFrame(
    {'Total Schools':[15], 
    'Total Students':[39170], 
    'Total Budget'  :[24649428],
    'Average Math Score':[78.99],
     'Average Reading Score':[81.88],
     '% Passing Reading':[85.81],
     '% Passing Math':[74.98],
     '%Overall Passing Rate':[80.43]
    }) 

# Print the dataframe 
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428               78.99   

   Average Reading Score  % Passing Reading  % Passing Math  \
0                  81.88              85.81           74.98   

   %Overall Passing Rate  
0                  80.43

#School Summary (Part2)

In [48]:
school_data_complete["passing_math"] = school_data_complete["math_score"] >= 70
school_data_complete["passing_reading"] = school_data_complete["reading_score"] >= 70

school_data_complete

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  \
0                 66          79          0  District  2917  1910635   
1                 94          61          0  District  2917  1910635   
2                 90          60          0  District  2917  1910635   
3                 67          58          0  District  2917  1910635   
4                 97          84          0  District  2917  1910635   
...              ...         ...        ...       ...   ...      ...   
39165             99          90         14   Charter  1635  1043130   
39166             95          70         14   Charter  1635  1043130   
39167             73          84         14   Charter  1635  1043130   
39168             99          90         14   Charter  1635  1043130   
39169             95          75         14   Charter  1635  1043130   

       passing_math  passing_reading  
0              True            False  
1             False             True  
2             False             True  
3             False            False  
4              True             True  
...             ...              ...  
39165          True             True  
39166          True             True  
39167          True             True  
39168          True             True  
39169          True             True  

[39170 rows x 13 columns]

In [65]:
#group by the different categorical columns
school_group = school_data_complete.groupby(["school_name"]).mean()

school_group["Per Student Budget"] = school_group["budget"]/school_group["size"]

school_group["% Passing Math"] = round(school_group["passing_math"]*100,2)

school_group["% Passing Reading"] = round(school_group["passing_reading"]*100,2)

school_group["% Overall Passing Rate"] = round(((school_group["passing_math"] 
                                                 + school_group["passing_reading"])/2)*100,2)


school_group.head(5)

Student ID  reading_score  math_score  School ID  \
school_name                                                              
Bailey High School       20358.5      81.033963   77.048432        7.0   
Cabrera High School      16941.5      83.975780   83.061895        6.0   
Figueroa High School      4391.0      81.158020   76.711767        1.0   
Ford High School         36165.0      80.746258   77.102592       13.0   
Griffin High School      12995.5      83.816757   83.351499        4.0   

                        size     budget  passing_math  passing_reading  \
school_name                                                              
Bailey High School    4976.0  3124928.0      0.666801         0.819333   
Cabrera High School   1858.0  1081356.0      0.941335         0.970398   
Figueroa High School  2949.0  1884411.0      0.659885         0.807392   
Ford High School      2739.0  1763916.0      0.683096         0.792990   
Griffin High School   1468.0   917500.0      0.933924         0.971390   

                      Per Student Budget  % Passing Math  % Passing Reading  \
school_name                                                                   
Bailey High School                 628.0           66.68              81.93   
Cabrera High School                582.0           94.13              97.04   
Figueroa High School               639.0           65.99              80.74   
Ford High School                   644.0           68.31              79.30   
Griffin High School                625.0           93.39              97.14   

                      % Overall Passing Rate  
school_name                                   
Bailey High School                     74.31  
Cabrera High School                    95.59  
Figueroa High School                   73.36  
Ford High School                       73.80  
Griffin High School                    95.27

Top Performing Schools (By Passing Rate)

In [72]:
top_five_schools = school_group.sort_values(["% Overall Passing Rate"], ascending=False)
top_five_schools.head()

Student ID  reading_score  math_score  School ID    size  \
school_name                                                                     
Cabrera High School     16941.5      83.975780   83.061895        6.0  1858.0   
Thomas High School      38352.0      83.848930   83.418349       14.0  1635.0   
Griffin High School     12995.5      83.816757   83.351499        4.0  1468.0   
Pena High School        23754.5      84.044699   83.839917        9.0   962.0   
Wilson High School      14871.0      83.989488   83.274201        5.0  2283.0   

                        budget  passing_math  passing_reading  \
school_name                                                     
Cabrera High School  1081356.0      0.941335         0.970398   
Thomas High School   1043130.0      0.932722         0.973089   
Griffin High School   917500.0      0.933924         0.971390   
Pena High School      585858.0      0.945946         0.959459   
Wilson High School   1319574.0      0.938677         0.965396   

                     Per Student Budget  % Passing Math  % Passing Reading  \
school_name                                                                  
Cabrera High School               582.0           94.13              97.04   
Thomas High School                638.0           93.27              97.31   
Griffin High School               625.0           93.39              97.14   
Pena High School                  609.0           94.59              95.95   
Wilson High School                578.0           93.87              96.54   

                     % Overall Passing Rate  
school_name                                  
Cabrera High School                   95.59  
Thomas High School                    95.29  
Griffin High School                   95.27  
Pena High School                      95.27  
Wilson High School                    95.20

Bottom Performing Schools (By Passing Rate)

In [86]:
Bottom_five_schools = school_group.sort_values(["% Overall Passing Rate"], ascending=True)
Bottom_five_schools.head()

Student ID  reading_score  math_score  School ID  \
school_name                                                               
Rodriguez High School     28035.0      80.744686   76.842711       11.0   
Figueroa High School       4391.0      81.158020   76.711767        1.0   
Huang High School          1458.0      81.182722   76.629414        0.0   
Johnson High School       32415.0      80.966394   77.072464       12.0   
Ford High School          36165.0      80.746258   77.102592       13.0   

                         size     budget  passing_math  passing_reading  \
school_name                                                               
Rodriguez High School  3999.0  2547363.0      0.663666         0.802201   
Figueroa High School   2949.0  1884411.0      0.659885         0.807392   
Huang High School      2917.0  1910635.0      0.656839         0.813164   
Johnson High School    4761.0  3094650.0      0.660576         0.812224   
Ford High School       2739.0  1763916.0      0.683096         0.792990   

                       Per Student Budget  % Passing Math  % Passing Reading  \
school_name                                                                    
Rodriguez High School               637.0           66.37              80.22   
Figueroa High School                639.0           65.99              80.74   
Huang High School                   655.0           65.68              81.32   
Johnson High School                 650.0           66.06              81.22   
Ford High School                    644.0           68.31              79.30   

                       % Overall Passing Rate  
school_name                                    
Rodriguez High School                   73.29  
Figueroa High School                    73.36  
Huang High School                       73.50  
Johnson High School                     73.64  
Ford High School                        73.80

Math Scores by Grade

In [89]:
#Create a series for each grade. 
ninth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "9th"].groupby("school_name").mean()["math_score"]
tenth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "10th"].groupby("school_name").mean()["math_score"]
eleventh_grade= school_data_complete[school_data_complete["grade"] 
                                    == "11th"].groupby("school_name").mean()["math_score"]
twelfth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "12th"].groupby("school_name").mean()["math_score"]

In [95]:
#Combine series into DataFrame 
math_scores_by_grade_df = pd.DataFrame({"9th":ninth_grade, 
                                       "10th": tenth_grade,
                                       "11th": eleventh_grade,
                                       "12th": twelfth_grade})
math_scores_by_grade_df.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    81.303155  80.907183  80.945643  80.912451
Cabrera High School   83.676136  84.253219  83.788382  84.287958
Figueroa High School  81.198598  81.408912  80.640339  81.384863
Ford High School      80.632653  81.262712  80.403642  80.662338
Griffin High School   83.369193  83.706897  84.288089  84.013699

Reading Scores by Grade

In [92]:
#Create a series for each grade. 
ninth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "9th"].groupby("school_name").mean()["reading_score"]
tenth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "10th"].groupby("school_name").mean()["reading_score"]
eleventh_grade= school_data_complete[school_data_complete["grade"] 
                                    == "11th"].groupby("school_name").mean()["reading_score"]
twelfth_grade= school_data_complete[school_data_complete["grade"] 
                                    == "12th"].groupby("school_name").mean()["reading_score"]

In [94]:
#Combine series into DataFrame 
reading_scores_by_grade_df = pd.DataFrame({"9th":ninth_grade, 
                                       "10th": tenth_grade,
                                       "11th": eleventh_grade,
                                       "12th": twelfth_grade})
reading_scores_by_grade_df.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    81.303155  80.907183  80.945643  80.912451
Cabrera High School   83.676136  84.253219  83.788382  84.287958
Figueroa High School  81.198598  81.408912  80.640339  81.384863
Ford High School      80.632653  81.262712  80.403642  80.662338
Griffin High School   83.369193  83.706897  84.288089  84.013699

Scores by School Spending

In [167]:
# Sample bins. Feel free to create your own bins.
spending_bins = [0, 585, 615, 645, 675]
group_names = ["<$585", "$585-615", "$615-645", "$645-675"]


# Use 4 reasonable bins to group school spending. 
school_data_summary["Spending Ranges (Per Student)"] = pd.cut(school_data_summary["Per Student Budget"], 
                                                              spending_bins, labels=group_names)

school_spending_grouped = school_data_summary.groupby("Spending Ranges (Per Student)").mean()


del school_spending_grouped['School ID_y']
del school_spending_grouped['Student ID']
del school_spending_grouped['size_x']
del school_spending_grouped['size_y']
del school_spending_grouped['School ID_x']
del school_spending_grouped['budget_x']
del school_spending_grouped['budget_y']
del school_spending_grouped['passing_reading']
del school_spending_grouped['passing_math']
del school_spending_grouped['Per Student Budget']

school_spending_grouped.head()

reading_score  math_score  % Passing Math  \
Spending Ranges (Per Student)                                              
<$585                              83.933814   83.455399       93.460000   
$585-615                           83.885211   83.599686       94.230000   
$615-645                           81.891436   79.079225       75.668333   
$645-675                           81.027843   76.997210       66.163333   

                               % Passing Reading  % Overall Passing Rate  
Spending Ranges (Per Student)                                             
<$585                                  96.610000               95.035000  
$585-615                               95.900000               95.065000  
$615-645                               86.106667               80.886667  
$645-675                               81.133333               73.650000

Scores by School Size

In [185]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

# Use 4 reasonable bins to group school size. 
school_data_summary["School Size"] = pd.cut(school_data_summary["size_x"], size_bins, labels=group_names)
school_data_summary

#group by size_x
school_size_grouped = school_data_summary.groupby("School Size").mean() 
school_size_grouped

#Remove the unwanted columns as per the sample provided
del school_size_grouped['budget_x']
del school_size_grouped['budget_y']
del school_size_grouped['Per Student Budget']
del school_size_grouped['School ID_y']
del school_size_grouped['School ID_x']
del school_size_grouped['passing_math']
del school_size_grouped['passing_reading']
del school_size_grouped['Student ID']
del school_size_grouped['size_x']
del school_size_grouped['size_y']

#Display
school_size_grouped

reading_score  math_score  % Passing Math  \
School Size                                                     
Small (<1000)           83.929843   83.821598        93.55000   
Medium (1000-2000)      83.864438   83.374684        93.59800   
Large (2000-5000)       81.344493   77.746417        69.96375   

                    % Passing Reading  % Overall Passing Rate  
School Size                                                    
Small (<1000)                96.10000                94.82500  
Medium (1000-2000)           96.79000                95.19600  
Large (2000-5000)            82.76625                76.36375

Scores by School Type

In [186]:
school_type_grouped = school_data_summary.groupby("type").mean()

#Remove the unwanted columns as per the sample provided
del school_type_grouped['size_x']
del school_type_grouped['size_y']
del school_type_grouped['budget_x']
del school_type_grouped['budget_y']
del school_type_grouped['Per Student Budget']
del school_type_grouped['School ID_y']
del school_type_grouped['School ID_x']
del school_type_grouped['passing_math']
del school_type_grouped['passing_reading']
del school_type_grouped['Student ID']

school_type_grouped

reading_score  math_score  % Passing Math  % Passing Reading  \
type                                                                     
Charter       83.896421   83.473852       93.620000          96.586250   
District      80.966636   76.956733       66.548571          80.798571   

          % Overall Passing Rate  
type                              
Charter                95.103750  
District               73.672857

Analysis:
    1. One observable trend is that Charter schools have a higher math and reading averages compared to District schools. 
    2. Smaller size schools have a higher passing percentage compared to large schools. On average, small size schools(with less than 1000 students), students have an overall 94% passing rate.
    3. In addition, schools that spent more money per student, also had a higher overall passing percentage. 
   
 Based on the data, it can be said that students that attend smaller schools that invest more money per student will likely have a higher math and reading average. 